### PR TITLE
TF-1083: Drag email to mailbox

### PR DESCRIPTION
**Pull request by ManhNTX**
https://user-images.githubusercontent.com/107173849/196932611-ee0d22e7-e658-42a0-987c-da2b06ff9c0d.mov

Drag with search mailbox: 

https://user-images.githubusercontent.com/107173849/197111718-cea92a86-b29b-47d3-8b04-8fd63214da78.mov

Drag with search email: 

https://user-images.githubusercontent.com/107173849/197125054-9cd95eee-856a-4dc5-b949-f1a00faa2c14.mov

After merge request when searching email :

https://user-images.githubusercontent.com/107173849/197466128-5f3b12ea-0f2d-421b-8921-9fed4e91dba6.mov

### CODE REVIEW

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 1]
```diff
@@ -672,6 +704,9 @@ class MailboxDashBoardController extends ReloadableController {
       destinationMailboxId = success.destinationMailboxId;
       moveAction = success.moveAction;
       emailActionType = success.emailActionType;
+      if (Get.isRegistered<ThreadController>()){
+        Get.find<ThreadController>().refreshEmailChanges(currentEmailState: success.currentEmailState);
+      }
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:709]: Why dashboard need to care to the threadView? It quite not clean? Please raise the viewState in dashboard and ThreadView should listen to it

File: lib/features/mailbox/presentation/widgets/mailbox_search_tile_builder.dart [hunk 2]
```diff
@@ -55,12 +57,24 @@ class MailboxSearchTileBuilder {
...
         splashColor: Colors.transparent,
         highlightColor: Colors.transparent),
-      child: _buildMailboxItem()
+      child: DragTarget<List<PresentationEmail>>(
+        builder: (context, _, __,) {
+          return _buildMailboxItem();
+        },
+        onAccept: (emails) {
+          _onDragItemAccepted?.call(emails, _presentationMailbox);
+          print(emails.first.preview);
```
Comment (dab246) [inline lib/features/mailbox/presentation/widgets/mailbox_search_tile_builder.dart:75]: remove it. or use logger

File: lib/features/thread/presentation/thread_view.dart [hunk 3]
```diff
@@ -332,29 +332,87 @@ class ThreadView extends GetWidget<ThreadController> with AppLoaderMixin,
...
+            data: controller.listEmailDrag,
+            child: (EmailTileBuilder(
+              context,
+              listPresentationEmail[index],
+              controller.mailboxDashBoardController.currentSelectMode.value,
+              controller.searchController.searchState.value.searchStatus,
+              controller.searchQuery,
+              mailboxCurrent: controller.searchController.isSearchEmailRunning
+                ? listPresentationEmail[index].findMailboxContain(
+                controller.mailboxDashBoardController.mapMailboxById)
```
Comment (dab246) [inline lib/features/thread/presentation/thread_view.dart:346]: add indent

**Comment by chibenwa** [comment]
ABSOLUTELY LOVE IT

**Comment by chibenwa** [comment]
Can I grag and drop from "Email Search" too?

**Comment by chibenwa** [comment]
Can I drag and drop when search filters of current mailbox are active (EG `important` ?)

**Comment by ManhNTX** [comment]
> Can I drag and drop when search filters of current mailbox are active


> Can I drag and drop when search filters of current mailbox are active (EG important ?)

Nice case. I updated evidence video for this case. Please check this !!! @chibenwa 

https://user-images.githubusercontent.com/107173849/197125313-1322abdc-84b0-407d-b8b1-71ef9312c47a.mov

**Comment by sherlockvn** [comment]
Look nice!. Can you show behavior when user drop to a destination, but the email can't be there

File: lib/features/email/domain/model/move_to_mailbox_request.dart [hunk 4]
```diff
@@ -1,32 +1,32 @@
...
 import 'package:jmap_dart_client/jmap/mail/email/email.dart';
 import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
 import 'package:model/model.dart';
 import 'package:tmail_ui_user/features/email/domain/model/move_action.dart';
 
 class MoveToMailboxRequest with EquatableMixin {
 
-  final List<EmailId> emailIds;
-  final MailboxId currentMailboxId;
+  final Map<MailboxId,List<EmailId>> mapEmailIdsByMailboxId;
```
Comment (hoangdat) [inline lib/features/email/domain/model/move_to_mailbox_request.dart:11]: how about `destinationMailboxes`

File: lib/features/email/data/network/email_api.dart [hunk 5]
```diff
@@ -302,31 +306,60 @@ class EmailAPI {
...
+        .execute();
+
+      Future.sync(() async {
+        final listUpdated = currentSetEmailInvocations
+          .map((currentInvocation) => response.parse(currentInvocation.methodCallId, SetEmailResponse.deserialize))
+          .map((e) => e!.updated!.keys).toList();
+
+        for (final listEmailId in listUpdated) {
+          List<EmailId> listEmailIdRequest = List.empty();
+          for (var itemEmailIdMoveRequest in moveRequest.currentMailboxes.values) {
```
Comment (dab246) [inline lib/features/email/data/network/email_api.dart:349]: Is there any other way without using 2 `for` loops here?

**Review by dab246 — APPROVED** [review]
/lgtm

File: lib/features/email/data/network/email_api.dart [hunk 6]
```diff
@@ -302,33 +306,61 @@ class EmailAPI {
...
-      .build()
-      .execute();
+    var start = 0;
+    var end = 0;
 
-    final setEmailResponse = response.parse<SetEmailResponse>(
-        setEmailInvocation.methodCallId,
-        SetEmailResponse.deserialize);
+    final List<EmailId> listEmailIdResult = List.empty(growable: true);
+    final listCurrentMailboxesEntries = moveRequest.currentMailboxes.entries.toList();
```
Comment (hoangdat) [inline lib/features/email/data/network/email_api.dart:320]: how about if an email belong to more than 1 mailboxes?

**Review by hoangdat — APPROVED** [review]

**Pull request merged**